File: Lintcode 069 - Binary Tree Level Order Traversal.java
Lintcode 069 - Binary Tree Level Order Traversal.java
Given a binary tree, return the level order traversal of its nodes' values. (ie, from left to right, level by level).
Example
Given binary tree {3,9,20,#,#,15,7},
    3
   / \
  9  20
    /  \
   15   7
return its level order traversal as:
[
  [3],
  [9,20],
  [15,7]
]

/**
 * Definition of TreeNode:
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left, right;
 *     public TreeNode(int val) {
 *         this.val = val;
 *         this.left = this.right = null;
 *     }
 * }
 */
 
public class Solution {
    public ArrayList<ArrayList<Integer>> levelOrder(TreeNode root) {
        // write your code here
        ArrayList<ArrayList<Integer>> result = new ArrayList<ArrayList<Integer>>();
        //用一个queue存储所有的Node。
        if(root == null) {
            return result;
        }
        
        Queue<TreeNode> queue = new LinkedList<TreeNode>();
        queue.offer(root);
        
        while(!queue.isEmpty()){
            ArrayList<Integer> level = new ArrayList<Integer>();
            //后面有poll操作，所以queue.size()会变，这里先存下来
            int size = queue.size(); 
            for(int i = 0; i < size; i++){
                TreeNode currNode = queue.poll();
                level.add(currNode.val);
                if (currNode.left != null) {
                    queue.offer(currNode.left);	
                }
                if (currNode.right != null) {
                    queue.offer(currNode.right);
                }
            }
            result.add(level);
        }
        return result;
    }
}

File: Lintcode 069 - Binary Tree Level Order Traversal.cs
using System;
using System.Collections;

namespace CSharp
{
    class TreeNode {
        public int val;
        public TreeNode left;
        public TreeNode right;
        public TreeNode (int val)
        {
            this.val = val;
            this.left = this.right = null;
        }
    }
    class Program
    {

        static void Main(string[] args)
        {

            TreeNode node1 = new TreeNode(1);
            TreeNode node2 = new TreeNode(2);
            TreeNode node3 = new TreeNode(3);
            TreeNode node4 = new TreeNode(4);
            TreeNode node5 = new TreeNode(5);
        
            node1.left = node2;
            node1.right = node3;
            node2.left = node4;
            node2.right = node5;

            ArrayList result = new ArrayList();
            Queue queue = new Queue();
        }

        static ArrayList LevelOrderTraversal(ArrayList result, Queue queue, TreeNode root)
        {
            if(root == null){
                return result;
            }

            queue.Enqueue(root);

            while(queue.Count > 0)
            {
                ArrayList level = new ArrayList();
                int size = queue.Count;

                for(int i = 0; i < size; i++)
                {
                    TreeNode currNode = queue.Dequeue();
                    level.Add(currNode);

                    if(currNode.left != null)
                    {
                        queue.Enqueue(currNode.left);
                    }

                    if(currNode.right != null)
                    {
                        queue.Enqueue(currNode.right);
                    }
                }

                result.Add(level);
            }

            return result;
        }

    }
}
